pkg/network/ebpf/c/protocols/postgres/decoding.h

```diff
@@ -240,19 +262,39 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
         // the message tag. So we need to add 1 to the message length to jump over the entire message.
         pktbuf_advance(pkt, header.message_len + 1);
     }
+    iteration_value->total_msg_count += messages_count;
 
     if (found_command_complete) {
+        handle_command_complete(&conn_tuple, transaction);
+
+        // save messages counter in bucket.
+        if (iteration_value->total_msg_count < PG_KERNEL_MSG_COUNT_FIRST_BUCKET) {
+            __sync_fetch_and_add(&pg_msg_counts->msg_count_buckets[0], 1);
+        } else {
+            __u8 bucket_idx = ((__u8)(iteration_value->total_msg_count - PG_KERNEL_MSG_COUNT_FIRST_BUCKET) / PG_KERNEL_MSG_COUNT_BUCKET_SIZE) + 1;
+            bucket_idx = (bucket_idx >= PG_KERNEL_MSG_COUNT_NUM_BUCKETS) ? (PG_KERNEL_MSG_COUNT_NUM_BUCKETS - 1) : bucket_idx;
+            __sync_fetch_and_add(&pg_msg_counts->msg_count_buckets[bucket_idx], 1);
+        }
         return 0;
     }
-    // We didn't find a command complete message, so we need to continue processing the packet.
-    // We save the current data offset and increment the iteration counter.
-    iteration_value->iteration += 1;
-    iteration_value->data_off = pktbuf_data_offset(pkt);
 
-    // If the maximum number of tail calls has been reached, we can skip invoking the next tail call.
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
+    if (iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)) {
+        // reached max messages, add counter and stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->reached_max_messages, 1);
+        return 0;
```

Comment:
you can reach to both `POSTGRES_MAX_TOTAL_MESSAGES` and have a fragmented read.
I don't think the return here is beneficial

Reply:
with current implementation conditions `(iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)) `and `(read_result == READ_FRAGMENTED)` may not happen simultaneously, because fragmentation is detected at the beginning of TCP packet

Reply:
I've removed `return` under this condition